dali/kernels/test/scatter_gather_test.cc

```diff
@@ -55,14 +56,33 @@ TEST(ScatterGather, Coalesce) {
   EXPECT_EQ(in, ref);
 }
 
-TEST(ScatterGather, Copy) {
+template <typename T>
+class ScatterGatherTest : public testing::Test {
+ public:
+  void Run(ScatterGatherCPU &sg, cudaStream_t, bool reset, ScatterGatherBase::Method,
+           ThreadPool &tp) {
+    sg.Run(tp, reset);
+  }
+
+  void Run(ScatterGatherGPU &sg, cudaStream_t stream, bool reset, ScatterGatherBase::Method method,
+           ThreadPool &) {
+    sg.Run(stream, reset, method);
+  }
+};
+
+TYPED_TEST_SUITE_P(ScatterGatherTest);
+
+TYPED_TEST_P(ScatterGatherTest, Copy) {
   const size_t max_l = 1024;
   std::vector<char> in(1<<20);
   std::vector<char> out(1<<20);
   unsigned seed = 42;
 
-  auto in_ptr = kernels::memory::alloc_unique<char>(AllocType::GPU, in.size());
-  auto out_ptr = kernels::memory::alloc_unique<char>(AllocType::GPU, out.size());
+  AllocType alloc =
+      std::is_same<TypeParam, ScatterGatherCPU>::value ? AllocType::Host : AllocType::GPU;
+
+  auto in_ptr = kernels::memory::alloc_unique<char>(alloc, in.size());
+  auto out_ptr = kernels::memory::alloc_unique<char>(alloc, out.size());
```

Comment:
I think that so much of this test is guarded with if-else that we'd be better off having two tests - one for GPU, one for CPU. Note that `kernels::memory` is slated for removal and there will be no run-time selection of memory kind!

Reply:
Moved to generic functions.